File: 1112-find-words-that-can-be-formed-by-characters/find-words-that-can-be-formed-by-characters.java
class Solution {
    int[] map(String word) {
        int[] ans = new int[26];
        for(char ch : word.toCharArray()) {
            ans[ch-'a']++;
        }
        return ans;
    }
    public int countCharacters(String[] words, String chars) {
        int ans = 0;
        int[] count = map(chars);
        for(String word : words) {
            int[] curr = map(word);
            boolean flag = true;
            for(int i=0; i<26; i++) {
                if(count[i] < curr[i]) {
                    flag = false;
                    break;
                }
            }
            if(flag) ans += word.length();
        }
        return ans;
    }
}

File: 1112-find-words-that-can-be-formed-by-characters/find-words-that-can-be-formed-by-characters.js
/**
 * @param {string[]} words
 * @param {string} chars
 * @return {number}
 */
var countCharacters = function(words, chars) {
    let alpha = 'abcdefghijklmnopqrstuvwxyz';
    let count = function(word) {
        let map = new Map();
        for(let c of alpha) map.set(c, 0);
        for(let ch of word) {
            map.set(ch, map.get(ch)+1);
        }
        return map;
    }
    
    let dict = count(chars);
    let ans = 0;
    for(let word of words) {
        let curr = count(word);
        let flag = true;
        for(let i of alpha) {
            if(curr.get(i) > dict.get(i)) {
                flag = false;
                break;
            }
        }
        if(flag) ans += word.length;
    }
    return ans;
};